oem/firstvoices/android/app/src/main/java/com/firstvoices/keyboards/SystemKeyboard.java

```diff
@@ -150,6 +150,10 @@ public void onStartInput(EditorInfo attribute, boolean restarting) {
             KMManager.setBannerOptions(mayPredict);
           }
         }
+
+        // Determine special handling for ENTER key
+        KMManager.setEnterMode(attribute.imeOptions, inputType);
+
         // User switched to a new input field so we should extract the text from input field
```

Comment:
Though you are only adding a couple lines to the code here, it would be nice to break up the functionality of `onStartInput` into several methods. It is about 50 lines long and is checking several conditions with discrete tasks.  Rather than having blocks of code with nice comments, it would be better to extract these into small well-named methods. Even if the methods are not used anywhere else, it will be more readable and unit-testable.

Reply:
Yep - that's a good suggestion (maybe can do some in early 19.0). 

A lot of the Android code could use refactoring (SystemKeyboard and the main culprit KMManager).